Content.Server/GameTicking/Rules/ZombieRuleSystem.cs

```diff
@@ -0,0 +1,309 @@
+
+    private void OnEntityZombified(EntityZombifiedEvent ev)
+    {
+        if (!Enabled)
+            return;
+        CheckRoundEnd(ev.Target);
+    }
+
+    /// <summary>
+    ///     The big kahoona function for checking if the round is gonna end
+    /// </summary>
+    /// <param name="target">depending on this uid, we should care about the round ending</param>
+    private void CheckRoundEnd(EntityUid target)
+    {
+        //we only care about players, not monkeys and such.
+        if (!HasComp<HumanoidAppearanceComponent>(target))
+            return;
+
+        var percent = GetInfectedPercentage(out var num);
+
+        if (num.Count == 1) //only one left
+            _popup.PopupEntity(Loc.GetString("zombie-alone"), num[0], Filter.Entities(num[0]));
+        if (percent >= 1) //oops, all zombies
+            _roundEndSystem.EndRound();
+    }
+
+    private void OnStartAttempt(RoundStartAttemptEvent ev)
+    {
+        if (!Enabled)
+            return;
+
+        var minPlayers = _cfg.GetCVar(CCVars.ZombieMinPlayers);
+        if (!ev.Forced && ev.Players.Length < minPlayers)
+        {
+            _chatManager.DispatchServerAnnouncement(Loc.GetString("zombie-not-enough-ready-players", ("readyPlayersCount", ev.Players.Length), ("minimumPlayers", minPlayers)));
+            ev.Cancel();
+            return;
+        }
+
+        if (ev.Players.Length == 0)
+        {
+            _chatManager.DispatchServerAnnouncement(Loc.GetString("zombie-no-one-ready"));
+            ev.Cancel();
+            return;
+        }
+    }
+
+    public override void Started(GameRuleConfiguration configuration)
+    {
+        //this technically will run twice with zombies on roundstart, but it doesn't matter because it fails instantly
+        InfectInitialPlayers();
+    }
+
+    public override void Ended(GameRuleConfiguration configuration) { }
+
+    private void OnZombifySelf(EntityUid uid, ZombifyOnDeathComponent component, ZombifySelfActionEvent args)
+    {
+        _zombify.ZombifyEntity(uid);
+
+        var action = new InstantAction(_prototypeManager.Index<InstantActionPrototype>(ZombifySelfActionPrototype));
+        _action.RemoveAction(uid, action);
+    }
+
+    private FixedPoint2 GetInfectedPercentage(out List<EntityUid> livingHumans)
+    {
+        var allPlayers = EntityQuery<HumanoidAppearanceComponent, MobStateComponent>();
+        var totalPlayers = new List<EntityUid>();
+        var livingZombies = new List<EntityUid>();
+        livingHumans = new();
+        foreach (var ent in allPlayers)
+        {
+            if (ent.Item2.IsAlive())
+            {
+                totalPlayers.Add(ent.Item2.Owner);
+
+                if (HasComp<ZombieComponent>(ent.Item1.Owner))
+                    livingZombies.Add(ent.Item2.Owner);
+                else
+                    livingHumans.Add(ent.Item2.Owner);
+            }
+        }
+        return (FixedPoint2) livingZombies.Count / (FixedPoint2) totalPlayers.Count;
+    }
```

Comment:
why is the percentage fixedpoint2 and not just float? it wasn't really intended for this. if its to round, then you should just use float and round for displaying. might also need to use `MathHelper.CloseTo` for the percentage checks if it doesn't work cause of float nonsense